### PR TITLE
Improve performance cache hit ratio and add integration tests

### PR DESCRIPTION
**Pull request by medb**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/PerformanceCachingGoogleCloudStorage.java [hunk 1]
```diff
@@ -112,9 +113,15 @@ public void deleteObjects(List<StorageResourceId> resourceIds) throws IOExceptio
   public List<GoogleCloudStorageItemInfo> listObjectInfo(
       String bucketName, String objectNamePrefix, ListObjectOptions listOptions)
       throws IOException {
-    List<GoogleCloudStorageItemInfo> result;
-
-    result = super.listObjectInfo(bucketName, objectNamePrefix, listOptions);
+    if (listOptions.getMaxResults() == 1 && listOptions.isIncludePrefix()) {
```
[the remark this answers is not in the document]
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/PerformanceCachingGoogleCloudStorage.java:116]: Ack.

[the remark this answers is not in the document]
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/PerformanceCachingGoogleCloudStorage.java:116]: This is an edge case in GCSIO implementation we use lists request with limit 1 as a hack for directory get request check, that's why we can special-case caching for it - we do not return cached result for any other list requests.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/PerformanceCachingGoogleCloudStorage.java [hunk 2]
```diff
@@ -140,18 +147,39 @@ public GoogleCloudStorageItemInfo getItemInfo(StorageResourceId resourceId) thro
...
+          directoryName == null
+              ? new StorageResourceId(bucketName)
+              : new StorageResourceId(bucketName, directoryName))) {
+        listObjectInfoPage(
+            bucketName, directoryName, ListObjectOptions.DEFAULT, /* pageToken= */ null);
+
+        item = cache.getItem(resourceId);
+
+        // If item is not in the cache but directory item is, then item does not exist.
+        if (item == null
```
[the remark this answers is not in the document]
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/PerformanceCachingGoogleCloudStorage.java:177]: I wanted to separate retrieval result from cache (this `if`) from initializing of cache with list request (next `if`).

**Review by hongyegong — APPROVED** [review]

**Pull request merged**